### PR TITLE
Adapt samples to sync API and require pre-allocated output buffers

### LINKED ISSUE #57: Adapt to sync API and require pre-allocated output buffers
Implement the changes introduced by https://github.com/webmachinelearning/webnn/pull/174

The changes include:
> 1. Change `MLGraphBuilder.build` and `MLGraph.compute` to sync API.
> 2. Change `MLGraph.compute` to require pre-allocated output buffers.
> 3. Simplify the the input / output resources binding by `MLResource`. Leave `MLInput` only for dynamic input shape case.

The webnn-polyfill PR https://github.com/webmachinelearning/webnn-polyfill/pull/85 is ready to merge.

Please adapt to that change before we merge that PR which will break the samples. Thanks!

### PR DESCRIPTION
**Pull request by Honry**
Fixed #57

### CODE REVIEW

**Comment by Honry** [comment]
@huningxin, I've verified all the samples with your PR locally, PTAL, thanks!

File: code/samples/dynamic_shape.js [hunk 1]
```diff
@@ -1,28 +1,29 @@
...
 const a = builder.input('a', descA);
 const descB = {type: 'float32', dimensions: [4, -1]};
 const b = builder.input('b', descB);
 const c = builder.matmul(a, b);
+const graph = builder.build({'c': c});
 
-const graph = await builder.build({c});
-
-async function compute(shapeA, shapeB) {
+function allocateAndCompute(shapeA, shapeB, shapeC) {
```
Comment (huningxin) [inline code/samples/dynamic_shape.js:12]: just name it as `compute`

File: code/samples/matmul.js [hunk 2]
```diff
@@ -10,9 +10,10 @@ const bufferB = new Float32Array(sizeOfShape(descB.dimensions)).fill(0.5);
...
 const c = builder.matmul(a, b);
 
-const graph = await builder.build({c});
+const graph = builder.build({c});
 const bufferA = new Float32Array(sizeOfShape(descA.dimensions)).fill(0.5);
-const inputs = {'a': {data: bufferA}};
-const outputs = await graph.compute(inputs);
-console.log(`shape: [${outputs.c.dimensions}]`);
-console.log(`values: ${outputs.c.data}`);
+const bufferC = new Float32Array(9);
```
Comment (huningxin) [inline code/samples/matmul.js:15]: `new Float32Array(sizeOfShape([3, 3]))`

File: image_classification/mobilenet_nhwc.js [hunk 3]
```diff
@@ -131,9 +131,11 @@ export class MobileNetV2Nhwc {
     }
   }
 
-  async compute(inputBuffer) {
-    const inputs = {input: {data: inputBuffer}};
-    const outputs = await this.graph_.compute(inputs);
-    return outputs;
+  compute(inputBuffer) {
```
Comment (huningxin) [inline image_classification/mobilenet_nhwc.js:134]: ditto.

File: image_classification/resnet101v2_nhwc.js [hunk 4]
```diff
@@ -180,9 +180,11 @@ export class ResNet101V2Nhwc {
     }
   }
 
-  async compute(inputBuffer) {
-    const inputs = {input: {data: inputBuffer}};
-    const outputs = await this.graph_.compute(inputs);
-    return outputs;
+  compute(inputBuffer) {
```
Comment (huningxin) [inline image_classification/resnet101v2_nhwc.js:183]: ditto.

File: image_classification/mobilenet_nchw.js [hunk 5]
```diff
@@ -132,9 +132,11 @@ export class MobileNetV2Nchw {
     }
   }
 
-  async compute(inputBuffer) {
-    const inputs = {input: {data: inputBuffer}};
-    const outputs = await this.graph_.compute(inputs);
-    return outputs;
+  compute(inputBuffer) {
+    const inputs = {'input': inputBuffer};
+    const outputBuffer = new Float32Array(1000);
```
Comment (huningxin) [inline image_classification/mobilenet_nchw.js:137]: probably you can pre-allocate the output buffer and reuse it for every compute.

File: image_classification/resnet50v2_nchw.js [hunk 6]
```diff
@@ -161,9 +161,11 @@ export class ResNet50V2Nchw {
     }
   }
 
-  async compute(inputBuffer) {
-    const inputs = {input: {data: inputBuffer}};
-    const outputs = await this.graph_.compute(inputs);
-    return outputs;
+  compute(inputBuffer) {
+    const inputs = {'input': inputBuffer};
+    const outputBuffer = new Float32Array(1000);
```
Comment (huningxin) [inline image_classification/resnet50v2_nchw.js:166]: ditto.

File: image_classification/squeezenet_nchw.js [hunk 7]
```diff
@@ -77,9 +77,11 @@ export class SqueezeNetNchw {
     }
   }
 
-  async compute(inputBuffer) {
-    const inputs = {input: {data: inputBuffer}};
-    const outputs = await this.graph_.compute(inputs);
-    return outputs;
+  compute(inputBuffer) {
+    const inputs = {'input': inputBuffer};
+    const outputBuffer = new Float32Array(1000);
```
Comment (huningxin) [inline image_classification/squeezenet_nchw.js:82]: ditto.

File: image_classification/squeezenet_nhwc.js [hunk 8]
```diff
@@ -85,9 +85,11 @@ export class SqueezeNetNhwc {
     }
   }
 
-  async compute(inputBuffer) {
-    const inputs = {input: {data: inputBuffer}};
-    const outputs = await this.graph_.compute(inputs);
-    return outputs;
+  compute(inputBuffer) {
+    const inputs = {'input': inputBuffer};
+    const outputBuffer = new Float32Array(1001);
```
Comment (huningxin) [inline image_classification/squeezenet_nhwc.js:90]: ditto.

File: lenet/lenet.js [hunk 9]
```diff
@@ -114,13 +114,15 @@ export class LeNet {
...
+    this.graph_ = this.builder_.build({'output': outputOperand});
   }
 
-  async predict(inputBuffer) {
-    const inputs = {input: {data: inputBuffer}};
-    const outputs = await this.graph_.compute(inputs);
-    return outputs.output.data;
+  predict(inputBuffer) {
+    const inputs = {'input': inputBuffer};
+    const outputBuffer = new Float32Array(10);
```
Comment (huningxin) [inline lenet/lenet.js:123]: ditto.

File: object_detection/ssd_mobilenetv1_nchw.js [hunk 10]
```diff
@@ -260,9 +259,15 @@ ${nameArray[1]}_BatchNorm_batchnorm`;
     }
   }
 
-  async compute(inputBuffer) {
-    const inputs = {input: {data: inputBuffer}};
-    const outputs = await this.graph_.compute(inputs);
+  compute(inputBuffer) {
+    const inputs = {'input': inputBuffer};
+    const boxesBuffer = new Float32Array(sizeOfShape([1, 1917, 1, 4]));
+    const scoresBuffer = new Float32Array(sizeOfShape([1, 1917, 91]));
```
Comment (huningxin) [inline object_detection/ssd_mobilenetv1_nchw.js:265]: pre-allocate the output buffers and reuse them.

File: object_detection/ssd_mobilenetv1_nhwc.js [hunk 11]
```diff
@@ -239,9 +238,15 @@ ${nameArray[1]}_BatchNorm_batchnorm`;
     }
   }
 
-  async compute(inputBuffer) {
-    const inputs = {input: {data: inputBuffer}};
-    const outputs = await this.graph_.compute(inputs);
+  compute(inputBuffer) {
+    const inputs = {'input': inputBuffer};
+    const boxesBuffer = new Float32Array(sizeOfShape([1, 1917, 1, 4]));
+    const scoresBuffer = new Float32Array(sizeOfShape([1, 1917, 91]));
```
Comment (huningxin) [inline object_detection/ssd_mobilenetv1_nhwc.js:244]: ditto.

File: object_detection/tiny_yolov2_nchw.js [hunk 12]
```diff
@@ -103,9 +103,11 @@ export class TinyYoloV2Nchw {
     }
   }
 
-  async compute(inputBuffer) {
-    const inputs = {input: {data: inputBuffer}};
-    const outputs = await this.graph_.compute(inputs);
+  compute(inputBuffer) {
+    const inputs = {'input': inputBuffer};
+    const outputBuffer = new Float32Array(sizeOfShape([1, 125, 13, 13]));
```
Comment (huningxin) [inline object_detection/tiny_yolov2_nchw.js:108]: ditto.

File: object_detection/tiny_yolov2_nhwc.js [hunk 13]
```diff
@@ -84,9 +84,11 @@ export class TinyYoloV2Nhwc {
     }
   }
 
-  async compute(inputBuffer) {
-    const inputs = {input: {data: inputBuffer}};
-    const outputs = await this.graph_.compute(inputs);
+  compute(inputBuffer) {
+    const inputs = {'input': inputBuffer};
+    const outputBuffer = new Float32Array(sizeOfShape([1, 13, 13, 125]));
```
Comment (huningxin) [inline object_detection/tiny_yolov2_nhwc.js:89]: ditto.

File: style_transfer/fast_style_transfer_net.js [hunk 14]
```diff
@@ -179,9 +179,12 @@ export class FastStyleTransferNet {
...
   }
 
-  async compute(inputBuffer) {
-    const inputs = {input: {data: inputBuffer}};
-    const outputs = await this.graph_.compute(inputs);
-    return outputs;
+  compute(inputBuffer) {
+    const inputs = {'input': inputBuffer};
+    const outputShape = [1, 3, 540, 540];
+    const outputBuffer = new Float32Array(sizeOfShape(outputShape));
```
Comment (huningxin) [inline style_transfer/fast_style_transfer_net.js:185]: ditto.

**Review by huningxin — COMMENTED** [review]
Thanks @Honry , left some comments, please take a look.

**Comment by Honry** [comment]
@huningxin, thanks for your comments! Fixed in new commits, PTAL again, thanks!

**Review by huningxin — APPROVED** [review]
LGTM, thanks @Honry !

**Pull request merged**